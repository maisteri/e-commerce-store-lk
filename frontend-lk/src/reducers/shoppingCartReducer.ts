import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { ShoppingCartItemId, ShoppingCartItem, ProductId } from '../types'
import { AppThunk } from '../store'
import cartService from '../services/cart'

const initialState: ShoppingCartItem[] = []

const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    append(state, action: PayloadAction<ShoppingCartItem>) {
      state.push(action.payload)
      console.log('appended')
      return state
    },
    setCart(_state, action: PayloadAction<ShoppingCartItem[]>) {
      console.log('set')
      return action.payload
    },
    remove(state, action: PayloadAction<ShoppingCartItemId>) {
      console.log('removed')
      return state.filter((item) => item.id !== action.payload)
    },
  },
})

export const addItemToCart = (productId: ProductId): AppThunk => {
  return async (dispatch, getState) => {
    const { shoppingCart } = getState()
    const cartItem = await cartService.addItem(productId)
    const itemAlreadyInCart = shoppingCart.find(
      (item) => item.productId === productId
    )
    if (itemAlreadyInCart) {
      const updatedShoppingCart = shoppingCart.map((item) =>
        item.productId === productId
          ? { ...item, quantity: item.quantity + 1 }
          : item
      )
      return dispatch(setCart(updatedShoppingCart))
    }
    dispatch(append(cartItem))
  }
}

export const initiateCart = (): AppThunk => {
  return async (dispatch) => {
    const products = await cartService.listItems()
    dispatch(setCart(products))
  }
}

export const modifyItemQuantity = (
  id: ShoppingCartItemId,
  quantity: number
): AppThunk => {
  return async (dispatch, getState) => {
    const { shoppingCart } = getState()
    const modifiedItem = await cartService.modifyItemQuantity(id, quantity)
    const updatedShoppingCart = shoppingCart.map((item) =>
      item.id === id ? modifiedItem : item
    )
    dispatch(setCart(updatedShoppingCart))
  }
}

export const deleteItem = (id: ShoppingCartItemId): AppThunk => {
  return async (dispatch) => {
    await cartService.deleteItem(id)
    dispatch(remove(id))
  }
}

export const { append, remove, setCart } = cartSlice.actions

export default cartSlice.reducer

