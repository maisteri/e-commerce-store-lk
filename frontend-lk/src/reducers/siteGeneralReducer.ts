import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { InitialGeneralState, NotificationIf, Product } from '../types'
import { AppThunk } from '../store'
import productService from '../services/product'
import { DISABLE_NOTIFICATION } from '../constants'

const initialState: InitialGeneralState = {
  sideDrawerOpen: false,
  categories: [],
  searchFilter: '',
  categorySelected: '',
  products: [],
  notification: {
    message: '',
    severity: 'success',
  },
}

const generalSlice = createSlice({
  name: 'general',
  initialState,
  reducers: {
    setSideDrawerOpen(state, action: PayloadAction<boolean>) {
      state.sideDrawerOpen = action.payload
    },
    setCategories(state, action: PayloadAction<Array<string>>) {
      state.categories = action.payload
    },
    setSelectedCategory(state, action: PayloadAction<string>) {
      state.categorySelected = action.payload
    },
    setSearchFilter(state, action: PayloadAction<string>) {
      state.searchFilter = action.payload
    },
    setProducts(state, action: PayloadAction<Product[]>) {
      state.products = action.payload
    },
    setNotification(state, action: PayloadAction<NotificationIf>) {
      state.notification = action.payload
    },
  },
})

export const notify = (notification: NotificationIf): AppThunk => {
  return async (dispatch) => {
    dispatch(setNotification(notification))
    setTimeout(() => {
      dispatch(setNotification(DISABLE_NOTIFICATION))
    }, 3000)
  }
}

export const initiateCategories = (): AppThunk => {
  return async (dispatch) => {
    const categories = await productService.getAllCategories()
    dispatch(setCategories(categories))
  }
}

export const initiateProducts = (): AppThunk => {
  return async (dispatch) => {
    const products = await productService.getAll({ category: '', search: '' })
    dispatch(setProducts(products))
  }
}

export const getProductsByCategory = (category: string): AppThunk => {
  return async (dispatch) => {
    const products = await productService.getAll({ category, search: '' })
    dispatch(setProducts(products))
  }
}

export const getProductsByFilter = (search: string): AppThunk => {
  return async (dispatch) => {
    const products = await productService.getAll({ category: '', search })
    dispatch(setProducts(products))
  }
}

export const {
  setSideDrawerOpen,
  setCategories,
  setSelectedCategory,
  setSearchFilter,
  setProducts,
  setNotification,
} = generalSlice.actions

export default generalSlice.reducer
