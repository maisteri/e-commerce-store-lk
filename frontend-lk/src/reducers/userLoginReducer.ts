import { PayloadAction, createSlice } from '@reduxjs/toolkit'
import loginService from '../services/login'
import { Credentials, UserData } from '../types'
import { AppThunk } from '../store'

const initialState: UserData = { token: '', name: '' }

const userSlice = createSlice({
  name: 'user',
  initialState: initialState,
  reducers: {
    setUser(_state, action: PayloadAction<UserData>) {
      return action.payload
    },
  },
})

export const initiateUser = (): AppThunk => {
  return (dispatch) => {
    const loggedUserJSON = window.localStorage.getItem('loggedLKAppUser')
    if (loggedUserJSON) {
      const user = JSON.parse(loggedUserJSON)
      dispatch(setUser(user))
    }
  }
}

export const loginUser = (credentials: Credentials): AppThunk => {
  return async (dispatch) => {
    const user = await loginService.login(credentials)
    window.localStorage.setItem('loggedLKAppUser', JSON.stringify(user))
    dispatch(setUser(user))
  }
}

export const logoutUser = (): AppThunk => {
  return async (dispatch) => {
    window.localStorage.removeItem('loggedLKAppUser')
    dispatch(setUser(initialState))
  }
}

export const { setUser } = userSlice.actions
export default userSlice.reducer
