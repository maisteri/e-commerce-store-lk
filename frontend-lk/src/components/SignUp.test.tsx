// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import SignUp from '../components/SignUp'

const navigate = jest.fn()
jest.mock('react-router-dom', () => ({
  useNavigate: () => navigate,
}))

const dispatch = jest.fn()
jest.mock('../hooks', () => ({
  useAppDispatch: () => dispatch,
}))

jest.mock('../reducers/userLoginReducer', () => {
  return {
    createNewUserAndLogin: jest.fn(),
  }
})
import { createNewUserAndLogin } from '../reducers/userLoginReducer'

const newUser = {
  name: 'testFirstName testLastName',
  password: 'testPassword',
  username: 'testEmail',
}

describe('<SignUp />', () => {
  it('renders correctly', () => {
    const { container } = render(<SignUp />)
    expect(container).toMatchSnapshot()
  })

  it('triggers dispatch and navigation on signup. Renders updated page after button push.', async () => {
    render(<SignUp />)

    const user = userEvent.setup()
    const firstNameInput = screen.getByLabelText(/First Name/)
    const lastNameInput = screen.getByLabelText(/Last Name/)
    const emailInput = screen.getByLabelText(/Email/)
    const passwordInput = screen.getByLabelText(/Password/)
    const signUpButton = screen.getByText('Sign Up')

    await user.type(firstNameInput, 'testFirstName')
    await user.type(lastNameInput, 'testLastName')
    await user.type(emailInput, 'testEmail')
    await user.type(passwordInput, 'testPassword')

    await user.click(signUpButton)

    expect(dispatch.mock.calls).toHaveLength(1)

    expect(createNewUserAndLogin.mock.calls).toHaveLength(1)
    expect(createNewUserAndLogin.mock.calls[0][0]).toEqual(newUser)
  })
})
