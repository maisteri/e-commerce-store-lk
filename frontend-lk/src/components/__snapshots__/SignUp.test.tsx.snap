// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders correctly 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-binzgt"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1999axt-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LockOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign up
      </h1>
      <form
        class="MuiBox-root css-178yklu"
        novalidate=""
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="firstName"
                id="firstName-label"
              >
                First Name
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="given-name"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="firstName"
                  name="firstName"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      First Name
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="lastName"
                id="lastName-label"
              >
                Last Name
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="family-name"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="lastName"
                  name="lastName"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Last Name
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="email"
                id="email-label"
              >
                Email Address
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="email"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="email"
                  name="email"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Email Address
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="new-password"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Password
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <label
              class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
            >
              <span
                class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              >
                <input
                  class="PrivateSwitchBase-input css-1m9pwf3"
                  data-indeterminate="false"
                  type="checkbox"
                  value="allowExtraEmails"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CheckBoxOutlineBlankIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
              >
                I want to receive inspiration, marketing promotions and updates via email.
              </span>
            </label>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Sign Up
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <a
              class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-101ca9i-MuiTypography-root-MuiLink-root"
              href="#"
            >
              Already have an account? Sign in
            </a>
          </div>
        </div>
      </form>
    </div>
  </main>
</div>
`;
